### PR TITLE
Allows modules to be objects in app definitions

### PR DESCRIPTION
**Pull request by ibolmo**
I ran into an edge case where I wanted to define various portions of the app definition in modules.

E.g. authentication:

```js
// ./authentication.js
export const test = () => {};
export const oauth2Config = {};
oauth2Config.authorizeUrl = {...};
```

in index.js

```js
import * as authentication from './authentication';

export default {
     authentication,
};
```

This works out of the box, but testing was breaking due to the `isPlainObject` check. I've added a plain module with an empty index.js to grab the constructor used by the runtime. I'll add a test, but wanted to run this by the team first.

### CODE REVIEW

**Comment by ibolmo** [comment]
added @bryanhelmig since it was suggested, but also cc: @BrunoBernardino @bcooksey who might have better context if this might cause 🔥

**Review by bryanhelmig — COMMENTED** [review]
Oh interesting -- I had no idea modules had different constructors... but it makes sense because _javascript_! 😛

File: src/tools/data.js [hunk 1]
```diff
@@ -2,9 +2,10 @@
 
 const _ = require('lodash');
 const memoize = require('./memoize');
+const plainModule = require('./plain');
 
 const isPlainObj = (o) => {
```
Comment (BrunoBernardino) [inline src/tools/data.js:7]: Can we do keep `isPlainObj` and add a `isPlainModule`, and then have the check for `isPlainObj || isPlainModule`? I think that's more clear than making an assumption and changing what `isPlainObj` really means.

**Review by BrunoBernardino — CHANGES_REQUESTED** [review]
Awesome, didn't know about this limitation as well, thank you so much for this!

I've made a comment asking for a change in the code to increase readability.

Additionally, if possible and not too complicated, could you also add a test for it?

File: src/tools/data.js (hunk 1 above)
Comment (ibolmo) [reply src/tools/data.js:7]: Looks like `isPlainObj` is used in another `serializeApp` via the `recurseCleanFuncs`. From my understanding, we'll want to permit the `isPlainObj || isPlainModule` check there as well but really I think we _should_ augment `isPlainObj`. The `o.constructor` check is really unforgiving. 

Let me know, if you really think this should be separate and if the recurseCleanFuncs should have the `|| isPlainModule`

Comment (BrunoBernardino) [reply src/tools/data.js:7]: You have the most context for it, so I'm definitely fine with keeping it like this.

**Review by BrunoBernardino — APPROVED** [review]

**Comment by ibolmo** [comment]
While (painfully) trying to recreate this locally I discovered this was a problem on my end:

**Notice the `lib/index` (the transpiled version) was used (incorrectly)**

```js
/* eslint-env jest */
import { createAppTester, tools } from 'zapier-platform-core';
import replace from 'replace-in-file';

tools.env.inject();

import App from 'lib/index';

const appTester = createAppTester(App);


test('should refresh auth', async () => {
  const result = await appTester(App.authentication.oauth2Config.refreshAccessToken, {
    authData: {
      refresh_token: process.env.REFRESH_TOKEN,
    },
  });
```

**Now see the fixed version (uses source ES7) [see my note below, if you're curious about the import]**

```js
/* eslint-env jest */
import { createAppTester, tools } from 'zapier-platform-core';
import replace from 'replace-in-file';

tools.env.inject();

import App from 'app';

const appTester = createAppTester(App);


test('should refresh auth', async () => {
  const result = await appTester(App.authentication.oauth2Config.refreshAccessToken, {
    authData: {
      refresh_token: process.env.REFRESH_TOKEN,
    },
  });
```

Closing since this is working now without the fix. 😓 


---

If you're curious about how I could do `import App from 'app';` here's my babelrc:

```json
{
  "presets": [
    ["env", {"targets": {"node": 6}}]
  ],
  "plugins": [
    "add-module-exports",
    "transform-object-rest-spread",
    ["module-resolver", {
      "root": ["./src"],
      "alias": {
        "app": "./src",
        "lib": "./lib",
        "test": "./test",
      }
    }]
  ]
}
```

The [module-resolver](https://github.com/tleunen/babel-plugin-module-resolver) is awesome.

**Comment by ibolmo** [comment]
So it turns out removing the `node_modules` and npm installing (thereby removing any local changes/fixes) brought back the error 😠 

I've spent quite a bit of time trying to reproduce it outside of the environment that this comes up (I'm using jest – which runs babel-node). 

At this point I'm thinking of merging without the additional test coverage. I'll keep trying to find a way to reproduce procedurally.

**Comment by ibolmo** [comment]
So.. quite a bit of turn of events.

It turns out that modules as objects is an "exotic" occurrence ([babeljs slack](https://babeljs.slack.com/archives/C0DFJT81H/p1506715554000324) if curious). Yet, the babel folks agree that modules should behave like objects. I've left a request in jest to review why the constructor is different, but at this point that's moot. I created a separate repo: https://github.com/ibolmo/namespaced-objects-conundrum that reproduces the problem and babel as well as jest will take a look.

For now, the lodash `isPlainObject` is working as expected for both `babel-node` and `jest`. A discussion arose if lodash will in the future make an exception for modules in the isPlainObject check, but we don't have to worry about that until Node.JS has native modules. Here's the line that will fail when native modules kick in: https://github.com/lodash/lodash/blob/4.17.4/lodash.js#L12059-L12060

Essentially they fall back to checking if the object's constructor is Object (like we did). But in native modules it'll be `[Function:  Module]`. 

Would be good to get your thoughts (@bryanhelmig  @BrunoBernardino) once more before proceeding.  

**My standing**: let's proceed using lodash' `isPlainObject` and if and when AWS gets native modules our test suite will throw a tantrum. Should we need to we can define our own checker like I did in the first commit. Likely, though, lodash (or the next gen lodash utility belt) will have implemented something.

**Comment by bryanhelmig** [comment]
I think we should probably roll a custom `isPlainObject` that handles both? If AWS updates the node version overnight that'll be a nasty outage! 😱

**Comment by ibolmo** [comment]
Alright we ended back to the original fix since I couldn't decouple the check for a new separate `isModuleObject`. I've planted the seed at lodash to consider working the problem, but going to default to action here. 

Should the spec decide to drop the `constructor` then it'll return `false` which is what we'd expect a plain object to be. Should the object be a module then the constructor should be the same. Jest has the curious behavior of  -- some how -- having a different constructor, but I've tested this separately (hard to do this in a mocha environment). 

Going to go ahead and merge.
